File: src/style.rs
use std::collections::HashMap;
use std::rc::Rc;

use crate::css;
use crate::css::Stylesheet;
use crate::dom::{Node, NodeKind, NodeRef};

/// Computed style property values.
#[derive(Debug, Clone)]
pub struct Style {
    pub properties: HashMap<String, String>,
}

impl Style {
    pub fn new() -> Self {
        Style {
            properties: HashMap::new(),
        }
    }

    /// Get a property value, or a default.
    pub fn get(&self, name: &str) -> &str {
        self.properties
            .get(name)
            .map(|s| s.as_str())
            .unwrap_or("")
    }

    /// Parse a length value (e.g., "16px" -> 16.0).
    pub fn length(&self, name: &str, default: f32) -> f32 {
        let val = self.get(name);
        if val.is_empty() {
            return default;
        }
        let val = val.trim();
        if let Some(px) = val.strip_suffix("px") {
            px.trim().parse().unwrap_or(default)
        } else if let Some(pct) = val.strip_suffix('%') {
            // Percentages are resolved later by the layout
            pct.trim().parse().unwrap_or(default)
        } else {
            val.parse().unwrap_or(default)
        }
    }

    /// Parse a length with percentage support (returns (px, is_percentage)).
    pub fn length_pct(&self, name: &str, default: f32) -> LengthValue {
        let val = self.get(name);
        if val.is_empty() {
            return LengthValue::Px(default);
        }
        let val = val.trim();
        if let Some(px) = val.strip_suffix("px") {
            LengthValue::Px(px.trim().parse().unwrap_or(default))
        } else if let Some(pct) = val.strip_suffix('%') {
            LengthValue::Pct(pct.trim().parse().unwrap_or(0.0))
        } else {
            // Try raw number as px
            LengthValue::Px(val.parse().unwrap_or(default))
        }
    }

    /// Parse a color value like "red", "#ff0000", "rgb(255,0,0)".
    pub fn color(&self, name: &str, default: (u8, u8, u8, u8)) -> (u8, u8, u8, u8) {
        let val = self.get(name);
        if val.is_empty() {
            return default;
        }
        parse_color(val).unwrap_or(default)
    }

    /// Get display property.
    pub fn display(&self) -> Display {
        match self.get("display") {
            "none" => Display::None,
            "inline" => Display::Inline,
            "inline-block" => Display::InlineBlock,
            "flex" => Display::Flex,
            "block" | _ => Display::Block,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Display {
    None,
    Block,
    Inline,
    InlineBlock,
    Flex,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum LengthValue {
    Px(f32),
    Pct(f32),
}

impl Default for Style {
    fn default() -> Self {
        Self::new()
    }
}

/// Parse a CSS color string.
pub fn parse_color(s: &str) -> Option<(u8, u8, u8, u8)> {
    let s = s.trim();
    if s.starts_with('#') {
        let hex = &s[1..];
        match hex.len() {
            3 => {
                let r = u8::from_str_radix(&hex[0..1], 16).ok()? * 17;
                let g = u8::from_str_radix(&hex[1..2], 16).ok()? * 17;
                let b = u8::from_str_radix(&hex[2..3], 16).ok()? * 17;
                Some((r, g, b, 255))
            }
            6 => {
                let r = u8::from_str_radix(&hex[0..2], 16).ok()?;
                let g = u8::from_str_radix(&hex[2..4], 16).ok()?;
                let b = u8::from_str_radix(&hex[4..6], 16).ok()?;
                Some((r, g, b, 255))
            }
            8 => {
                let r = u8::from_str_radix(&hex[0..2], 16).ok()?;
                let g = u8::from_str_radix(&hex[2..4], 16).ok()?;
                let b = u8::from_str_radix(&hex[4..6], 16).ok()?;
                let a = u8::from_str_radix(&hex[6..8], 16).ok()?;
                Some((r, g, b, a))
            }
            _ => None,
        }
    } else if s.starts_with("rgb(") {
        let inner = s.trim_start_matches("rgb(").trim_end_matches(')');
        let parts: Vec<&str> = inner.split(',').collect();
        if parts.len() == 3 {
            let r = parts[0].trim().parse().ok()?;
            let g = parts[1].trim().parse().ok()?;
            let b = parts[2].trim().parse().ok()?;
            Some((r, g, b, 255))
        } else {
            None
        }
    } else if s.starts_with("rgba(") {
        let inner = s.trim_start_matches("rgba(").trim_end_matches(')');
        let parts: Vec<&str> = inner.split(',').collect();
        if parts.len() == 4 {
            let r = parts[0].trim().parse().ok()?;
            let g = parts[1].trim().parse().ok()?;
            let b = parts[2].trim().parse().ok()?;
            let a = (parts[3].trim().parse::<f32>().ok()? * 255.0) as u8;
            Some((r, g, b, a))
        } else {
            None
        }
    } else {
        // Named colors (common subset)
        let named: HashMap<&str, (u8, u8, u8)> = [
            ("black", (0, 0, 0)),
            ("white", (255, 255, 255)),
            ("red", (255, 0, 0)),
            ("green", (0, 128, 0)),
            ("blue", (0, 0, 255)),
            ("yellow", (255, 255, 0)),
            ("orange", (255, 165, 0)),
            ("purple", (128, 0, 128)),
            ("gray", (128, 128, 128)),
            ("grey", (128, 128, 128)),
            ("pink", (255, 192, 203)),
            ("brown", (165, 42, 42)),
            ("navy", (0, 0, 128)),
            ("teal", (0, 128, 128)),
            ("transparent", (0, 0, 0)),
        ]
        .iter()
        .cloned()
        .collect();
        named.get(s).map(|&(r, g, b)| {
            if s == "transparent" {
                (r, g, b, 0)
            } else {
                (r, g, b, 255)
            }
        })
    }
}

/// Build a stylesheet from a `<style>` tag content and optional inline styles.
pub fn collect_styles(doc: &NodeRef) -> Stylesheet {
    let mut css_text = String::new();
    collect_style_nodes(&doc.borrow(), &mut css_text);
    css::parse_css(&css_text)
}

fn collect_style_nodes(node: &Node, css: &mut String) {
    if let NodeKind::Element(e) = &node.kind {
        if e.tag == "style" {
            for child in &node.children {
                if let NodeKind::Text(t) = &child.borrow().kind {
                    css.push_str(t);
                }
            }
        }
    }
    for child in &node.children {
        collect_style_nodes(&child.borrow(), css);
    }
}

/// Compute the style for a single node from the stylesheet.
pub fn compute_style(node: &Node, stylesheet: &Stylesheet) -> Style {
    let mut style = Style::new();

    // Apply default styles for block elements
    let tag = node.tag_name();
    apply_defaults(tag, &mut style);

    // Apply matched rules from stylesheet
    let matched = css::match_rules(node, stylesheet);
    for rule in matched {
        for decl in &rule.declarations {
            style.properties.insert(decl.name.clone(), decl.value.clone());
        }
    }

    // Apply inline style attribute
    if let Some(data) = node.element_data() {
        if let Some(inline) = data.get_attr("style") {
            let inline_ss = css::parse_css(&format!("_ {{ {} }}", inline));
            for rule in &inline_ss.rules {
                for decl in &rule.declarations {
                    style.properties.insert(decl.name.clone(), decl.value.clone());
                }
            }
        }
    }

    style
}

fn apply_defaults(tag: Option<&str>, style: &mut Style) {
    match tag {
        Some("body") | Some("html") | Some("div") | Some("p")
        | Some("h1") | Some("h2") | Some("h3") | Some("h4") | Some("h5") | Some("h6")
        | Some("ul") | Some("ol") | Some("li") | Some("table") | Some("tr")
        | Some("td") | Some("th") | Some("header") | Some("footer") | Some("nav")
        | Some("section") | Some("article") | Some("main") | Some("aside") | Some("hr") => {
            style.properties.insert("display".to_string(), "block".to_string());
        }
        Some("span") | Some("a") | Some("b") | Some("i") | Some("strong") | Some("em") => {
            style.properties.insert("display".to_string(), "inline".to_string());
        }
        _ => {}
    }

    // Default font sizes for headings
    match tag {
        Some("h1") => { style.properties.insert("font-size".to_string(), "32px".to_string()); }
        Some("h2") => { style.properties.insert("font-size".to_string(), "24px".to_string()); }
        Some("h3") => { style.properties.insert("font-size".to_string(), "18px".to_string()); }
        Some("h4") => { style.properties.insert("font-size".to_string(), "16px".to_string()); }
        Some("h5") => { style.properties.insert("font-size".to_string(), "14px".to_string()); }
        Some("h6") => { style.properties.insert("font-size".to_string(), "12px".to_string()); }
        _ => {}
    }

    // Default font-weight for bold tags
    match tag {
        Some("b") | Some("strong") => {
            style.properties.insert("font-weight".to_string(), "bold".to_string());
        }
        _ => {}
    }
    // Default margin for body
    if tag == Some("body") {
        style.properties.insert("margin".to_string(), "8px".to_string());
    }
    // Default margin for paragraphs
    if tag == Some("p") {
        style.properties.insert("margin".to_string(), "16px 0".to_string());
    }
    // Default list padding
    if tag == Some("ul") || tag == Some("ol") {
        style.properties.insert("padding-left".to_string(), "40px".to_string());
    }
    // Default hr
    if tag == Some("hr") {
        style.properties.insert("border".to_string(), "1px solid black".to_string());
        style.properties.insert("margin".to_string(), "8px 0".to_string());
    }
}

/// Recursively compute styles for all nodes in the tree.
/// Returns a map keyed by node pointer address to avoid Hash/Eq issues with Rc<RefCell<Node>>.
pub fn compute_styles(doc: &NodeRef, stylesheet: &Stylesheet) -> HashMap<usize, Style> {
    let mut map = HashMap::new();
    compute_styles_recursive(doc, stylesheet, &mut map);
    map
}

fn node_key(node: &NodeRef) -> usize {
    Rc::as_ptr(node) as *const Node as usize
}

fn compute_styles_recursive(
    node: &NodeRef,
    stylesheet: &Stylesheet,
    map: &mut HashMap<usize, Style>,
) {
    let style = compute_style(&node.borrow(), stylesheet);
    map.insert(node_key(node), style);
    for child in &node.borrow().children {
        compute_styles_recursive(child, stylesheet, map);
    }
}

File: android/app/src/main/java/com/easyweb/app/engine/Style.kt
// b站：绝望彻彻底底的绝望 [android/app/src/main/java/com/easyweb/app/engine/Style.kt]

package com.easyweb.app.engine

import java.util.HashMap

// 长度值类型
sealed class LengthValue {
    data class Px(val value: Float) : LengthValue()
    data class Pct(val value: Float) : LengthValue()
}

// 显示类型
enum class Display {
    None, Block, Inline, InlineBlock, Flex
}

// 计算后的样式
class Style {
    val properties = HashMap<String, String>()

    // 获取属性值
    fun get(name: String): String = properties[name] ?: ""

    // 解析长度值
    fun length(name: String, default: Float): Float {
        val value = get(name)
        if (value.isEmpty()) return default
        val trimmed = value.trim()
        return if (trimmed.endsWith("px")) {
            trimmed.removeSuffix("px").trim().toFloatOrNull() ?: default
        } else {
            trimmed.toFloatOrNull() ?: default
        }
    }

    // 解析带百分比的长度值
    fun lengthPct(name: String, default: Float): LengthValue {
        val value = get(name)
        if (value.isEmpty()) return LengthValue.Px(default)
        val trimmed = value.trim()
        return if (trimmed.endsWith("px")) {
            LengthValue.Px(trimmed.removeSuffix("px").trim().toFloatOrNull() ?: default)
        } else if (trimmed.endsWith("%")) {
            LengthValue.Pct(trimmed.removeSuffix("%").trim().toFloatOrNull() ?: 0.0f)
        } else {
            LengthValue.Px(trimmed.toFloatOrNull() ?: default)
        }
    }

    // 解析颜色值
    fun color(name: String, default: Long): Long {
        val value = get(name)
        if (value.isEmpty()) return default
        return parseColor(value) ?: default
    }

    // 获取显示类型
    fun display(): Display {
        return when (get("display")) {
            "none" -> Display.None
            "inline" -> Display.Inline
            "inline-block" -> Display.InlineBlock
            "flex" -> Display.Flex
            else -> Display.Block
        }
    }
}

// 解析颜色字符串 返回ARGB格式的Long
// 颜色存储在Long中 低32位为ARGB
fun parseColor(s: String): Long? {
    val trimmed = s.trim()
    return when {
        trimmed.startsWith("#") -> {
            val hex = trimmed.substring(1)
            when (hex.length) {
                3 -> {
                    val r = hex.substring(0, 1).toIntOrNull(16)?.times(17) ?: return null
                    val g = hex.substring(1, 2).toIntOrNull(16)?.times(17) ?: return null
                    val b = hex.substring(2, 3).toIntOrNull(16)?.times(17) ?: return null
                    toArgb(r, g, b, 255)
                }
                6 -> {
                    val r = hex.substring(0, 2).toIntOrNull(16) ?: return null
                    val g = hex.substring(2, 4).toIntOrNull(16) ?: return null
                    val b = hex.substring(4, 6).toIntOrNull(16) ?: return null
                    toArgb(r, g, b, 255)
                }
                8 -> {
                    val r = hex.substring(0, 2).toIntOrNull(16) ?: return null
                    val g = hex.substring(2, 4).toIntOrNull(16) ?: return null
                    val b = hex.substring(4, 6).toIntOrNull(16) ?: return null
                    val a = hex.substring(6, 8).toIntOrNull(16) ?: return null
                    toArgb(r, g, b, a)
                }
                else -> null
            }
        }
        trimmed.startsWith("rgb(") -> {
            val inner = trimmed.removePrefix("rgb(").removeSuffix(")").trim()
            val parts = inner.split(",")
            if (parts.size == 3) {
                val r = parts[0].trim().toIntOrNull() ?: return null
                val g = parts[1].trim().toIntOrNull() ?: return null
                val b = parts[2].trim().toIntOrNull() ?: return null
                toArgb(r, g, b, 255)
            } else null
        }
        trimmed.startsWith("rgba(") -> {
            val inner = trimmed.removePrefix("rgba(").removeSuffix(")").trim()
            val parts = inner.split(",")
            if (parts.size == 4) {
                val r = parts[0].trim().toIntOrNull() ?: return null
                val g = parts[1].trim().toIntOrNull() ?: return null
                val b = parts[2].trim().toIntOrNull() ?: return null
                val a = (parts[3].trim().toFloatOrNull()?.times(255)?.toInt()) ?: return null
                toArgb(r, g, b, a)
            } else null
        }
        else -> {
            // 命名颜色
            val named = namedColors[trimmed.lowercase()]
            named?.let { (r, g, b) ->
                if (trimmed.lowercase() == "transparent") toArgb(r, g, b, 0)
                else toArgb(r, g, b, 255)
            }
        }
    }
}

// 将RGBA转换为ARGB long
fun toArgb(r: Int, g: Int, b: Int, a: Int): Long {
    return ((a.toLong() and 0xFF) shl 24) or
           ((r.toLong() and 0xFF) shl 16) or
           ((g.toLong() and 0xFF) shl 8) or
           (b.toLong() and 0xFF)
}

// 提取ARGB分量
fun alphaOf(color: Long): Int = ((color shr 24) and 0xFF).toInt()
fun redOf(color: Long): Int = ((color shr 16) and 0xFF).toInt()
fun greenOf(color: Long): Int = ((color shr 8) and 0xFF).toInt()
fun blueOf(color: Long): Int = (color and 0xFF).toInt()

// 常用命名颜色
private val namedColors = mapOf(
    "black" to Triple(0, 0, 0),
    "white" to Triple(255, 255, 255),
    "red" to Triple(255, 0, 0),
    "green" to Triple(0, 128, 0),
    "blue" to Triple(0, 0, 255),
    "yellow" to Triple(255, 255, 0),
    "orange" to Triple(255, 165, 0),
    "purple" to Triple(128, 0, 128),
    "gray" to Triple(128, 128, 128),
    "grey" to Triple(128, 128, 128),
    "pink" to Triple(255, 192, 203),
    "brown" to Triple(165, 42, 42),
    "navy" to Triple(0, 0, 128),
    "teal" to Triple(0, 128, 128),
    "transparent" to Triple(0, 0, 0)
)

// 从样式表中收集所有样式
fun collectStyles(doc: Node): Stylesheet {
    val cssText = StringBuilder()
    collectStyleNodes(doc, cssText)
    return CssParser.parse(cssText.toString())
}

// 递归收集style标签内容
private fun collectStyleNodes(node: Node, css: StringBuilder) {
    if (node.kind == NodeKind.Element && node.tagName() == "style") {
        for (child in node.children) {
            if (child.kind == NodeKind.Text && child.text != null) {
                css.append(child.text)
            }
        }
    }
    for (child in node.children) {
        collectStyleNodes(child, css)
    }
}

// 计算单个节点的样式
fun computeStyle(node: Node, stylesheet: Stylesheet): Style {
    val style = Style()

    // 应用默认样式
    applyDefaults(node.tagName(), style)

    // 应用匹配的规则
    val matched = matchRules(node, stylesheet)
    for (rule in matched) {
        for (decl in rule.declarations) {
            style.properties[decl.name] = decl.value
        }
    }

    // 应用内联样式
    val data = node.elementData
    if (data != null) {
        val inline = data.getAttr("style")
        if (inline != null) {
            val inlineSs = CssParser.parse("_ { $inline }")
            for (rule in inlineSs.rules) {
                for (decl in rule.declarations) {
                    style.properties[decl.name] = decl.value
                }
            }
        }
    }

    return style
}

// 应用默认样式
private fun applyDefaults(tag: String?, style: Style) {
    when (tag) {
        "body", "html", "div", "p", "h1", "h2", "h3", "h4", "h5", "h6",
        "ul", "ol", "li", "table", "tr", "td", "th",
        "header", "footer", "nav", "section", "article", "main", "aside", "hr" -> {
            style.properties["display"] = "block"
        }
        "span", "a", "b", "i", "strong", "em" -> {
            style.properties["display"] = "inline"
        }
    }

    // 标题默认字号
    when (tag) {
        "h1" -> style.properties["font-size"] = "32px"
        "h2" -> style.properties["font-size"] = "24px"
        "h3" -> style.properties["font-size"] = "18px"
        "h4" -> style.properties["font-size"] = "16px"
        "h5" -> style.properties["font-size"] = "14px"
        "h6" -> style.properties["font-size"] = "12px"
    }

    // 粗体标签
    when (tag) {
        "b", "strong" -> style.properties["font-weight"] = "bold"
    }

    // 默认边距
    if (tag == "body") {
        style.properties["margin"] = "8px"
    }
    if (tag == "p") {
        style.properties["margin"] = "16px 0"
    }
    if (tag == "ul" || tag == "ol") {
        style.properties["padding-left"] = "40px"
    }
    if (tag == "hr") {
        style.properties["border"] = "1px solid black"
        style.properties["margin"] = "8px 0"
    }
}

// 递归计算所有节点的样式
fun computeStyles(doc: Node, stylesheet: Stylesheet): Map<Node, Style> {
    val map = HashMap<Node, Style>()
    computeStylesRecursive(doc, stylesheet, map)
    return map
}

// 递归计算样式
private fun computeStylesRecursive(node: Node, stylesheet: Stylesheet, map: MutableMap<Node, Style>) {
    map[node] = computeStyle(node, stylesheet)
    for (child in node.children) {
        computeStylesRecursive(child, stylesheet, map)
    }
}